docs/source/models.rst

```diff
@@ -81,6 +81,98 @@ time_created
+
+Model Signature
+^^^^^^^^^^^^^^^
+Model signature defines schema of model input and output. Both model input and output can be
+described as a sequence of (optionally) named columns with type specified as one of the
+:py:class:`MLflow data types <mlflow.types.DataType>`. The signature is stored
+in JSON format in MLmodel file together with other model metadata. Model signatures are  recognized
+(and enforced) by standard MLflow model tools, such as when deploying python function as a REST API
+endpoint with mlflow model serve.
+
+Below is an example  of model signature for classification model build over well known iris dataset.
+The input is 4 named numeric columns, output is unnamed integer specifying the predicted class:
+
+.. code-block:: yaml
+
+  signature:
+      inputs: '[{"name": "sepal length (cm)", "type": "double"}, {"name": "sepal width
+        (cm)", "type": "double"}, {"name": "petal length (cm)", "type": "double"}, {"name":
+        "petal width (cm)", "type": "double"}]'
+      outputs: [{"type": "integer"}]
+
+Signature Enforcement
+~~~~~~~~~~~~~~~~~~~~~
+MLflow will enforce input schema when scoring models that include signature. The schema enforcement
+checks input column ordering and column types and will raise an exception if the input is not
+compatible. This enforcement is applied in MLflow before calling the underlying model
+implementation. Note that this enforcement only applies when using MLflow model deployment tools or
+when loading models as ``python_function``. In particular, it is not applied to models that are
+loaded in their native format (e.g. by calling mlflow.sklearn.load_model).
+
+Column Ordering Enforcement
+"""""""""""""""""""""""""""
+The input columns are checked against the model signature. If there are any missing columns,
+MLflow will raise an exception. Extra columns that were not declared in the signature will be
+ignored. If the input schema in the signature defines column names, column matching is done by name
+and the columns are reordered to match the signature. If the input schema does not have column
+names, matching is done by position (i.e. MLflow will only check the number of columns).
+
+Column Type Enforcement
+"""""""""""""""""""""""
+The input column types are checked against the signature. MLflow will perform safe type conversions
+if necessary. Generally, only upcasts (e.g. integer -> long or float -> double) are considered to be
+safe. If the types can not be made compatible, MLflow will raise an error.
+
+How To Log Models With Signature
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The model signature can be :py:func:`inferred <mlflow.models.infer_signature>`
+from datasets with valid model input (e.g. the training dataset with target
+column omitted) and valid model output (e.g. model predictions generated on
+the training dataset). To include signature with your model, add it to the appropriate log_model
```

Comment:
I think we can probably include more detail here about which argument is used for the signature in `log_model()` and what the type of its value should be. Perhaps some discussion of the [ModelSignature](https://mlflow.org/docs/latest/python_api/mlflow.models.html#mlflow.models.ModelSignature) object with a link to the docs would help.

I think it would be useful to discuss this "how to specify a signature in log_model()" part first and then discuss inferring the schema below (i.e., switch the order). Finally, would it be difficult to include a small example where the  signature object is constructed manually?

Reply:
Makes sense.